modules/aws/ssm.go

```diff
@@ -68,3 +75,141 @@ func NewSsmClientE(t testing.TestingT, region string) (*ssm.SSM, error) {
+	})
+
+	return err
+}
+
+// WaitForSsmInstance waits until the instance get registered to the SSM inventory.
+func WaitForSsmInstance(t testing.TestingT, awsRegion, instanceID string, timeout time.Duration) {
+	if err := WaitForSsmInstanceE(t, awsRegion, instanceID, timeout); err != nil {
+		t.Fatalf("failed to find %s in the SSM store: %s", instanceID, err.Error())
+	}
+}
+
+// CheckSsmCommand checks that you can run the given command on the given instance through AWS SSM. Returns stdout and stderr.
+func CheckSsmCommand(t testing.TestingT, awsRegion, instanceID, command string, timeout time.Duration) (string, string) {
+	stdout, stderr, err := CheckSsmCommandE(t, awsRegion, instanceID, command, timeout)
+	if err != nil {
+		t.Fatalf("failed to execute '%s' on %s (%v):]\n  stdout: %#v\n  stderr: %#v", command, instanceID, err, stdout, stderr)
+	}
+	return stdout, stderr
+}
+
+type result struct {
+	Stdout string
+	Stderr string
+}
+
+// CheckSsmCommandE checks that you can run the given command on the given instance through AWS SSM. Returns the stdout, stderr and an error if one occurs.
+func CheckSsmCommandE(t testing.TestingT, awsRegion, instanceID, command string, timeout time.Duration) (string, string, error) {
+	logger.Logf(t, "Running command %s on %s", command, instanceID)
+
+	timeBetweenRetries := 2 * time.Second
+	maxRetries := int(timeout.Seconds() / timeBetweenRetries.Seconds())
+
+	// Now that we know the instance in the SSM inventory, we can send the command
+	client, err := NewSsmClientE(t, awsRegion)
+	if err != nil {
+		return "", "", err
+	}
+	comment := "Terratest SSM"
+	documentName := "AWS-RunShellScript"
+	req, resp := client.SendCommandRequest(&ssm.SendCommandInput{
+		Comment:      &comment,
+		DocumentName: &documentName,
+		InstanceIds:  []*string{&instanceID},
+		Parameters: map[string][]*string{
+			"commands": []*string{&command},
+		},
+	})
+	if err := req.Send(); err != nil {
+		return "", "", err
+	}
+
+	// Wait for the result
+	description := "Waiting for the result of the command"
+	retryableErrors := map[string]string{
+		"InvocationDoesNotExist": "InvocationDoesNotExist",
+		"bad status: Pending":    "bad status: Pending",
+		"bad status: InProgress": "bad status: InProgress",
+		"bad status: Delayed":    "bad status: Delayed",
+	}
+	out, err := retry.DoWithRetryableErrorsE(t, description, retryableErrors, maxRetries, timeBetweenRetries, func() (string, error) {
+		client, err := NewSsmClientE(t, awsRegion)
+		if err != nil {
+			return "", err
+		}
+		req, resp := client.GetCommandInvocationRequest(&ssm.GetCommandInvocationInput{
+			CommandId:  resp.Command.CommandId,
+			InstanceId: &instanceID,
+		})
+		if err := req.Send(); err != nil {
+			return "", err
+		}
+
+		// Remove the SSM prefix from stderr
+		stdErr := *resp.StandardErrorContent
+		prefix := regexp.MustCompile(`/var/lib/amazon/ssm/.*/_script\.sh: line 1: `)
+		stdErr = prefix.ReplaceAllString(stdErr, "")
+
+		b, err := json.Marshal(result{
```

Comment:
Please use a full variable name and not one-character versions. After scrolling down a few lines, I couldn't remember what `b` was.

Reply:
Oh, also, I'm guessing you're creating a JSON string here so you can then return it from the retry function and unmarshall it back into a struct later... The reason the retry function returns a string is that Go doesn't support generics, so it was either `interface{}` and all our code casting, or `string`, and the function then works well in a few cases, and poorly in others. I think instead of marshalling to/from JSON, in this case, you may want to declare the return value variables outside the retry loop and set them to the values you want within the loop. You can then return `"", nil` to exit the loop.